feature: Add bdbVersion/spareCores/withClang/withDebug/withGui args

cbergqvist opened the pull request:
Let me know if you want me to break the changes into separate PRs!

Example usage:
```console
$ nix-shell --arg withClang true --arg withDebug true --arg withGui false --arg spareCores 2 --argstr bdbVersion db48
```

0xB10C commented:
Please rebase on master for CI.

I think we would want to document the argument strings in a table in README.md? Or other ideas?

Maybe also a clang CI job would be good? Maybe replacing the job matrix with a few defined combinations..

0xB10C commented:
I was so free and rebased this for the CI to run on it. I'll have a closer look once CI passes.

cbergqvist committed:
Add sections on installation and usage.

cbergqvist committed:
feature: Add withClang/withDebug/withGui args

cbergqvist committed:
improve: Jobs querying for Mac + add spareCores arg

Due to the authors limited nix-fu, we hand the responsibility of not subtracting too many cores to user.

cbergqvist committed:
feature: Berkeley DB support is now configurable up to a point

cbergqvist commented:
Did some minor updates + added content to README.md along the lines you specified. Let me know what you think.

Don't know if a Clang CI job is necessary. Is Clang not covered by Bitcoin Core CI?

0xB10C commented on shell.nix:
```diff
@@ -1,10 +1,26 @@
+  withDebug ? false,
+  withGui ? false,
+}:
 #pkgs.clangStdenv.mkDerivation {
 #  name = "libcxxStdenv";
 # clang_13
```
nit: Could drop this

0xB10C replied:
Done in #9

0xB10C reviewed (approved):
Looks good. Played around with it a bit and it seems to work well. Thank you!

Any particular reason for having both `db48` and `db5`? I assumed `db48` would be sufficient.

0xB10C commented:
> Don't know if a Clang CI job is necessary. Is Clang not covered by Bitcoin Core CI?

Clang is covered by the Core CI, a CI here would test that `nix-shell --arg withClang true` continues works fine. By default, were only testing GCC here and might not notice if something on the NixOS or Bitcoin Core side changes in regards to clang. Probably not super important though.

cbergqvist commented:
> Any particular reason for having both `db48` and `db5`? I assumed `db48` would be sufficient.

Just another "test vector" since the build system supports `--with-incompatible-bdb`.

cbergqvist commented:
> Clang is covered by the Core CI, a CI here would test that nix-shell --arg withClang true continues works fine.

I'm fine either way. Nice to have as long as whoever pays the CI bills is okay with it.
